internal/dhcp.go

```
@@ -0,0 +1,303 @@
+		log.Println(err)
+	}
+
+}
+
+func (udp *UDPConnection) Connect(ipaddr string, destport string) error {
+	var err error
+	udp.Connection, err = net.Dial("udp", ipaddr+":"+destport)
+	if err != nil {
+		log.Println("failed dial step ", err)
+		return err
+	}
+	return nil
+}
+
+func (udp *UDPConnection) Write(pkt Packet) error {
+	_, err := udp.Connection.Write(pkt)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (udp *UDPConnection) Close() error {
+	if udp.Connection == nil {
+		return errors.New("no connection to close")
+	}
+
+	err := udp.Connection.Close()
+	udp.Connection = nil
+	return err
+}
+
+func createReplyPacket(msgType MessageType, serverId net.IP, yIAddr net.IP, opitons []Option) Packet {
+	p := NewPacket(bootreply)
+	p.SetXId(IntToByteArray(10392900, 4))
+	p.SetFlags(IntToByteArray(32768, 2))
+	p.SetYIAddr(yIAddr)
+	p.SetGIAddr(net.ParseIP("0.0.0.0"))
+	x, _ := net.ParseMAC(clientmac)
+	p.SetCHAddr(x)
+	p.AddOption(OptionDHCPMessageType, []byte{byte(msgType)})
+	p.AddOption(OptionServerIdentifier, []byte(serverId))
+	for _, opt := range opitons {
+		p.AddOption(opt.Code, opt.Value)
+	}
+	p.PadToMinSize()
+	return p
+}
+
+func setDHCPOptions() []Option {
+	var opts []Option
+	addDHCPOption(&opts, OptionSubnetMask, subnetmask.To4())
+	addDHCPOption(&opts, OptionTimeOffset, IntToByteArray(0, 4))
+	addDHCPOption(&opts, OptionDomainNameServer, dns.To4())
+	addDHCPOption(&opts, OptionDomainName, []byte(domain))
+	addDHCPOption(&opts, OptionDefaultTTL, []byte{64})
+	addDHCPOption(&opts, OptionIPLeaseTime, IntToByteArray(86400, 4))
+	addDHCPOption(&opts, OptionRenewalTime, IntToByteArray(43200, 4))
+	addDHCPOption(&opts, OptionRebindingTime, IntToByteArray(75600, 4))
+	return opts
+}
+
+func addDHCPOption(array *[]Option, code OptionCode, value []byte) {
+	tmp := Option{Code: code, Value: value}
+	*array = append(*array, tmp)
+}
+
+func getIPV4Addr(ne NetworkEnumerator) string {
+	list, err := ne.Interfaces()
+	if err != nil {
+		panic(err)
```

Generally, we don't want to terminate the program unexpectedly. We should return the error to the caller, log it, and exit gracefully if it is something we can't recover from -- which is unlikely since what we'll probably want to do is try again.